### PR TITLE
Add support for `com.hazelcast:hazelcast:5.2.1`

### PR DESCRIPTION
**Pull request by linghengqian**
## What does this PR do?

- Fixes https://github.com/oracle/graalvm-reachability-metadata/issues/32  .
- Fixes https://github.com/oracle/graalvm-reachability-metadata/pull/122 .
- This is part of https://github.com/oracle/graalvm-reachability-metadata/issues/117 .

## Checklist before merging
- [x] I have properly formatted metadata files (see [CONTRIBUTING](/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md) document)
- [x] I have added thorough tests. (see [this](/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#Tests))

### CODE REVIEW

**Comment by linghengqian** [comment]
- I finally realized that the out of memory issue came from com.hazelcast.client.HazelcastClient, not Hazelcast Server, which is ridiculous. I feel like this is a bug in Hazelcast 5.2.1 and I need to adjust the target version of the PR.

File: tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (mhalbritter) [inline tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json:1]: This looks like that the Hazelcast support needs `javax.cache`. I guess without those files the tests won't run. If this is the case, these files should be provided as a separate PR to support the `javax.cache:cache-api` dependencies.

**Review by mhalbritter — CHANGES_REQUESTED** [review]

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -1,10 +1,10 @@
 [versions]
 # Project versions
-nativeBuildTools = "0.9.16"
```
[the remark this answers is not in the document]
Comment (linghengqian) [reply gradle/libs.versions.toml:3]: - This is essentially just to test what's going on in Github Actions, I'll roll back this part.

File: tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json (hunk 1 above)
Comment (linghengqian) [reply tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json:1]: - I did open it https://github.com/oracle/graalvm-reachability-metadata/issues/136 .
- However, I have no idea how to change the content of `serialization-config. json` of `javax.cache:cache-api:1.1.1` to the `graalvm reachability metadata conditional` form. I added that part manually. The error log of GraalVM gives little information about serialization.

**Comment by linghengqian** [comment]
- @Mhalbriter Hi, I want to know if you know why calling the method of `com.hazelcast.client.HazelcastClient` will cause the Github Actions device to crash? This branch is normally executed locally `./gradlew clean test -Pcoordinates=com.hazelcast:hazelcast:5.2.1`, but Github Actions will crash due to device restrictions. 
- I have no idea about this. Hazelcast `5.1.5`, `5.2.0`, `5.2.1` showed the same problem in Github Actions. I understand that this seems to be the problem of Hazelcast, but I can't find the relevant issue.
- image

File: tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json (hunk 1 above)
Comment (mhalbritter) [reply tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json:1]: Ah, thanks. I missed that.

Comment (linghengqian) [reply tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json:1]: > Would this help to make it conditional?
> 
> https://docs.oracle.com/en/graalvm/enterprise/22/docs/reference-manual/native-image/metadata/

- I've seen the docs before. The reason why I need to manually add this part of metadata is that when I use GraalVM Trace Agent, even if I configure the conditional mode for the `javax.cache.*` package, I cannot collect the GraalVM reachability metadata related to serialization. I have to turn on the standard mode to manually find the serialization metadata that I really need, which makes me don't know where to start. 

- The key point is that when the metadata related to serialization is missing, the error log of GraalVM CE 22.3 does not prompt what the context of the required class is, which is very different from the prompt when the metadata of reflection is missing.

Comment (vjovanov) [reply tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json:1]: OK, these both seem like bugs, thanks for pointing it out. Just to clarify, by context you mean the stack trace where serialization should happen? 

Would it be possible to open a [ticket](https://github.com/oracle/graal/issues) with a command to reproduce?

Comment (linghengqian) [reply tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json:1]: - If you think this is a bug, I will open a new issue at https://github.com/oracle/graal and provide a reproducible example. 

- So far, I have only found that the serialization metadata of `javax.cache.*` is required in the unit test of hazelcast. I am not sure whether this is related to https://github.com/oracle/graalvm-reachability-metadata/pull/169 #issuecomment-1378981023 . hazelcast's GraalVM reachability metadata differs significantly from other JSR-107 implementations, such as Ehcache and Caffeine.

**Comment by mhalbritter** [comment]
Hey, no idea why the Github Actions won't run. I have no broader experience with either Hazelcast nor Github Actions, so I'm afraid I can't really help on that.

**Comment by linghengqian** [comment]
> Hey, no idea why the Github Actions won't run. I have no broader experience with either Hazelcast nor Github Actions, so I'm afraid I can't really help on that.

- From this perspective, I don't think we should consider incorporating this PR at GraalVM Reachability Metadata 0.2.6. Of course, if a friend submits another hazelcast PR and gets merged, I'd be happy to try to add more single tests to the next release milestone.

**Comment by linghengqian** [comment]
- So according to the findings of @fniephaus in https://github.com/oracle/graalvm-reachability-metadata/pull/122#issuecomment-1381913698, I determined that Hazelcast has a completely different port performance in Github Actions than that of local devices. And this actually didn't print any errors. So I specified all ports in https://github.com/linghengqian/graalvm-trace-metadata-smoketest/commit/7f6a0ee59ab6ade4cf5a6348aabc2819d8993072 and put the processing into the current PR.

- At the same time, because Hazelcast is difficult to specify the port on JCache, I replaced `System.setProperty("hazelcast.jcache.provider.type", "client")` with `System.setProperty("hazelcast.jcache.provider.type", "member")`, which is missing part of the unit tests related to https://docs.hazelcast.com/hazelcast/5.2/jcache/providers#hide-nav, but I don't think this will affect.

- Due to a new discovery at https://github.com/oracle/graalvm-reachability-metadata/pull/170, the `org.junit.jupiter.api.Timeout` annotation cannot be used under nativeTest, so I introduced `org. awaitility:awaitility:4.2.0` to test the timeout behavior of HazelcastClient.

- Regarding the problem of `serialization-config.json` of `javax.cache:cache-api:1.1.1`, I misunderstood the meaning of Error Log before, and I have modified it to the GraalVM reachability metadata in the conditional style. I now know that this Log means that if and only if the `java.lang.Object` class is reachable, consider the `javax.cache.configuration.FactoryBuilder$SingletonFactory` entry. I am at https://github.com/linghengqian/graalvm-trace-metadata-smoketest/commit/80bc33ff5d532a4b1daa78030103abf0233ee2a6 fixes this, and addresses the current PR as well.
```shell
 JUnit Jupiter:HazelcastTest:testJCache()
    MethodSource [className = 'com.lingh.HazelcastTest', methodName = 'testJCache', methodParameterTypes = '']
    => com.oracle.svm.core.jdk.UnsupportedFeatureError: SerializationConstructorAccessor class not found for declaringClass: javax.cache.configuration.FactoryBuilder$SingletonFactory (targetConstructorClass: java.lang.Object). Usually adding javax.cache.configuration.FactoryBuilder$SingletonFactory to serialization-config.json fixes the problem.
       org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:89)
       org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.serialize.SerializationSupport.getSerializationConstructorAccessor(SerializationSupport.java:143)
       java.base@17.0.5/jdk.internal.reflect.MethodAccessorGenerator.generateSerializationConstructor(MethodAccessorGenerator.java:48)
       java.base@17.0.5/jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:463)
       java.base@17.0.5/jdk.internal.reflect.ReflectionFactory.newConstructorForSerialization(ReflectionFactory.java:455)
       java.base@17.0.5/java.io.ObjectStreamClass.getSerializableConstructor(ObjectStreamClass.java:1443)
       java.base@17.0.5/java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:412)
       java.base@17.0.5/java.io.ObjectStreamClass$2.run(ObjectStreamClass.java:384)
       java.base@17.0.5/java.security.AccessController.executePrivileged(AccessController.java:168)
       java.base@17.0.5/java.security.AccessController.doPrivileged(AccessController.java:318)
       [...]
```
- This PR is ready for review.

File: tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json (hunk 1 above)
Comment (linghengqian) [reply tests/src/com.hazelcast/hazelcast/5.2.1/src/test/resources/META-INF/native-image/javax.cache/cache-api/1.1.1/resource-config.json:1]: @vjovanov I misunderstood the Log, I dealt with this at https://github.com/oracle/graalvm-reachability-metadata/pull/169#issuecomment-1382704703, there doesn't seem to be anything to handle on the GraalVM side.

**Review by mhalbritter — APPROVED** [review]
Looks good to me.

**Comment by linghengqian** [comment]
- Since https://github.com/oracle/graalvm-reachability-metadata/pull/176 was merged, I removed the manually added `javax.cache:cache-api:1.1.1` GraalVM reachability metadata in this PR in the latest commit. 

- It seems that since this week, the current Git CI has encountered a problem. Every time a PR is merged, all unclosed PRs will be outdate.

**Review by dnestoro — APPROVED** [review]

**Comment by dnestoro** [comment]
> * Since [Add support for `javax.cache:cache-api:1.1.1` #176](https://github.com/oracle/graalvm-reachability-metadata/pull/176) was merged, I removed the manually added `javax.cache:cache-api:1.1.1` GraalVM reachability metadata in this PR in the latest commit.
> * It seems that since this week, the current Git CI has encountered a problem. Every time a PR is merged, all unclosed PRs will be outdate.

We changed some repo configuration so it shouldn't happen anymore. If that won't fix the problem, we will explore some other options
